docker/docker.ts

```diff
@@ -175,36 +180,76 @@ async function buildAndPushImageWorkerAsync(
+    }
 
-    // Use those then push the image.  Then just return the digest as the final result for our caller to use.
+    // Generate a name that uniquely will identify this built image.  This is similar in purpose to
+    // the name@digest form that can be normally be retrieved from a docker repository.  However,
+    // this tag doesn't require actually pushing the image, nor does it require communicating with
+    // some external system, making it suitable for unique identification, even during preview.
+    // This also means that if docker produces a new imageId, we'll get a new name here, ensuring that
+    // resources (like docker.Image and cloud.Service) will be appropriately replaced.
+    const uniqueTargetName = createTargetName(repositoryUrl, tag, imageId);
+
+    // Use those then push the image.  Then just return the unique target name. as the final result
+    // for our caller to use.
     if (!pulumi.runtime.isDryRun()) {
-        // Only push the image during an update, do not push during a preview, even if digest and url are available
-        // from a previous update.
+        // Only push the image during an update, do not push during a preview.
         await login();
 
         // Push the final image first, then push the stage images to use for caching.
-        await pushImageAsync(imageName, repositoryUrl, logResource);
+        await tagAndPushImageAsync(imageName, repositoryUrl, tag, imageId, logResource);
 
-        for (const stage of buildResult.stages) {
-            await pushImageAsync(
-                localStageImageName(imageName, stage), repositoryUrl, logResource, stage);
+        for (const stage of stages) {
+            await tagAndPushImageAsync(
+                localStageImageName(imageName, stage), repositoryUrl, tag, /*imageId:*/ undefined, logResource);
         }
     }
 
-    return buildResult.id;
+    return uniqueTargetName;
+}
+
+function localStageImageName(imageName: string, stage: string) {
+    return `${imageName}-${stage}`;
+}
+
+function getImageNameAndTag(baseImageName: string): { imageName: string, tag: string | undefined } {
+    const lastColon = baseImageName.lastIndexOf(":");
+    const imageName = lastColon < 0 ? baseImageName : baseImageName.substr(0, lastColon);
+    const tag = lastColon < 0 ? undefined : baseImageName.substr(lastColon + 1);
+
+    return  { imageName, tag };
+}
+
+function createTargetName(repositoryUrl: string, tag: string | undefined, imageId: string | undefined): string {
+    const pieces: string[] = [];
+    if (tag) {
+        pieces.push(tag);
+    }
+
+    if (imageId) {
+        pieces.push(imageId);
+    }
+
+    // A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits,
+    // underscores, periods and dashes. A tag name may not start with a period or a dash and may
+    // contain a maximum of 128 characters.
+    const fullTag = pieces.join("-").replace(/[^-_.a-zA-Z0-9]/, "").substr(0, 128);
```

Comment:
Do we have to do this?  If the user provided a tag - it's their fault if it isn't legal and I don't think we should try to "fix" it.  If the imageID from Docker is not legal to use here - we should probably do a more intentional transformation on it.  Or if we are paranoid that it's format may change arbitrarily - at least only do this on the imageId component?

Reply:
that's a fair point.  The major concern is simply the length.  If the caller provides a long tag length, and we jam on the hash, and we go past 128, then that's going to blowup.  But perhaps we just let that blowup in that case....